internal/api/v1/servicebindings.go

```diff
@@ -26,71 +31,90 @@ func (hc ServicebindingsController) Create(w http.ResponseWriter, r *http.Reques
+			return APIErrors{BadRequest(err)}
 		}
 	}
 
 	cluster, err := kubernetes.GetCluster()
 	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+		return APIErrors{InternalError(err)}
 	}
 
 	exists, err := organizations.Exists(cluster, org)
 	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+		return APIErrors{InternalError(err)}
 	}
 	if !exists {
-		return APIErrors{
-			NewAPIError(fmt.Sprintf("Organization '%s' does not exist", org), "", http.StatusNotFound),
-		}
+		return APIErrors{OrgIsNotKnown(org)}
 	}
 
 	app, err := application.Lookup(cluster, org, appName)
 	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+		return APIErrors{InternalError(err)}
 	}
 	if app == nil {
-		return APIErrors{
-			NewAPIError(fmt.Sprintf("application '%s' not found", appName), "", http.StatusNotFound),
-		}
+		return APIErrors{AppIsNotKnown(appName)}
 	}
 
-	service, err := services.Lookup(cluster, org, bindRequest.Name)
-	if err != nil && err.Error() == "service not found" {
-		return APIErrors{
-			NewAPIError(fmt.Sprintf("service '%s' not found", bindRequest.Name), "", http.StatusNotFound),
+	// From here on out we collect errors and warnings per
+	// service, to report as much as possible while also applying
+	// as much as possible. IOW even when errors are reported it
+	// is possible for some of the services to be properly bound.
+
+	var theServices interfaces.ServiceList
+	var theIssues APIErrors
+
+	for _, serviceName := range bindRequest.Names {
+		service, err := services.Lookup(cluster, org, serviceName)
+		if err != nil {
+			if err.Error() == "service not found" {
+				theIssues = append(theIssues, ServiceIsNotKnown(serviceName))
+				continue
+			}
+
+			return append(APIErrors{InternalError(err)}, theIssues...)
 		}
-	}
-	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+
+		theServices = append(theServices, service)
 	}
 
-	err = app.Bind(service)
-	if err != nil && err.Error() == "service already bound" {
-		return APIErrors{
-			NewAPIError(fmt.Sprintf("service '%s' already bound", bindRequest.Name), "", http.StatusConflict),
+	for _, service := range theServices {
+		err = app.Bind(service)
+		if err != nil {
+			if err.Error() == "service already bound" {
+				theIssues = append(theIssues, ServiceAlreadyBound(service.Name()))
+				continue
+			}
+
+			return append(APIErrors{InternalError(err)}, theIssues...)
 		}
 	}
-	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+
+	if len(theIssues) > 0 {
+		return theIssues
 	}
 
 	w.Header().Set("Content-Type", "application/json")
 	_, err = w.Write([]byte{})
 	if err != nil {
-		return APIErrors{NewAPIError(err.Error(), "", http.StatusInternalServerError)}
+		return append(APIErrors{InternalError(err)}, theIssues...)
```

Comment:
with the current code, there is no way for `theIssues` to be non-empty at this point (see line 110 above). Should we keep this just in case someone adds code between line 110 and 117 that populates `theIssues`?

Reply:
Checking...

Reply:
Right. I forgot to simplify that when I fixed an issue, i.e. adding the `return theIssues`. Will fix.

Reply:
Done with commit c86ec5ad0204734e53a780788412a84c284b10d6